### ISSUE TITLE
IRIS: Support for IRIS Sample Changer

### ISSUE THREAD

**Issue by kjwoodsISIS** [ISISComputingGroup/IBEX#1407]
As scientist working on IRIS, I need to control the sample changer, so that I can choose which sample is placed in the beam.

##### Acceptance Criteria
1. I can view which sample is currently in the beam (i.e. view which sample is currently selected).
1. I can change which sample is currently in the beam (i.e. change thje selected sample).
1. I can view and change the selected sample via the IBEX GUI.

##### Notes
1. The IRIS sample changer is closely based on the existing MERLIN sample changer.
1. Ideally we can re-use the existing sample changer IOC used on LARMOR (it is just named positions).  We might need a new OPI (show only one degree of freedom, rather than two?).  Check with instrument scientist.

**Comment by AdrianPotter** [comment]
@davidkeymer: Is there a reference for the sample changer available we can use to build the protocol?

**Comment by kjwoodsISIS** [comment]
The sample changer is driven by a set of motion control set points.  For more details see: https://github.com/ISISComputingGroup/IBEX/wiki/Motion-Control

**Comment by AdrianPotter** [comment]
Waiting on ticket #1099. The sample changer is controlled by a Mclennan Motion Controller. Once ticket #1099 is done, we just need an OPI specialised to Iris's requirements.

**Comment by AdrianPotter** [comment]
For the next dev who picks this up, if it's not me, consider seeing if David would like a hand with ticket 1099 in case he's tied up doing other things.

**Comment by kjwoodsISIS** [comment]
See notes on [McLennan](https://github.com/ISISComputingGroup/IBEX/wiki/IRIS-Instrument-Details#noteMcLennan) and 
[Sample Changer](https://github.com/ISISComputingGroup/IBEX/wiki/IRIS-Instrument-Details#noteSampleChanger) on the [IRIS Instrument Deatils](https://github.com/ISISComputingGroup/IBEX/wiki/IRIS-Instrument-Details) page.  The notes include links to a photograph and diagram (sketch, with dimensions) of the Sample Changer.

**Comment by AdrianPotter** [comment]
I've created a linear sample changer based on the Galil motor controller.  The Iris sample changer is intended to use a McLennan. This is covered under tickets #1099 and #1100. Marking as impeded until those tickets are ready

**Comment by kjwoodsISIS** [comment]
@AdrianPotter: I don't think this story is impeded.  The sample changer should, in principle, be able to work with any motor (it should just be a matter of pointing it at the right motor).  I would prefer to see this story tested and completed, rather than wait for the McLennan tickets to be completed.
We can also create a new issue to configure the IRIS sample changer to use McLennan motors.

**Comment by AdrianPotter** [comment]
Ok, I'll move it to review and it can be tested with the Galil example config

**Comment by AdrianPotter** [comment]
SampleChanger support module changes: https://github.com/ISISComputingGroup/EPICS-sampleChanger/pull/2

**Comment by AdrianPotter** [comment]
EPICS submodule updates: https://github.com/ISISComputingGroup/EPICS/pull/37

**Comment by AdrianPotter** [comment]
GUI: https://github.com/ISISComputingGroup/ibex_gui/pull/297

**Comment by KathrynBaker** [comment]
Moving back to rework (so that it can be picked up when there is time) as I found the following:
* Whenever I reload the sample changer configuration (using the buttons in the GUI) I get the following message: `NDW1032:KVLB23:LKUP:LSAMPLE:_SPLITPOSNS incorrect output type FTVM != STRING` reported as an error
* The EPICS PR now has merge errors
* I'm not sure about the inclusion of the config example, I actually went to the extent of adding them in by hand, as I have a list of other bits set up and didn't want to have to back them up and adding new ones was potentially less confusing for me, but I can see where others would find that more useful, but the example doesn't feel clean enough for storing the source code, as it has the old sample changer in it, and a whole load of files that are standard to the configurations directory which feels unnecessary in a code repo. 

The first one is the one that potentially needs to be fixed, the second is probably just a merge and OK it all by hand, and the third is almost a non-issue, so as long as 1 and 2 are sorted this can be merged.

**Comment by AdrianPotter** [comment]
For the final point, I left both the existing sample changer and the linear sample changer in the example so each is demonstrated in the example. Given that there is a single support module, for both linear and non-linear, it seemed sensible that both should be demonstrated in the example.

2 is easy (touch wood).

Will have a look at 1. I did see it, but I believe it occurs with the current sample changer too, so originally considered it outside the scope of the ticket.

**Comment by KathrynBaker** [comment]
It’s not the two sample changers I was overly concerned about (both of them in place made sense), but the inclusion of the _base aspects as well, and all the other axes and motion set point items, the full set of galil.cmds, things like beamstop.txt, the inclusion of linsample.txt (this file is created from other xml files), including a globals.txt. It means a lot of files were created that I’m not sure should be kept in this way, after all, including a synoptic, and the full configuration, means that changes to how those are managed would put the content of the repo out of date on things which don’t relate to the repo and the function of the item it is focussed on anyway. Stating a few assumptions, and including example files is one thing, including a complete configuration feels like overkill, and could potentially cause confusion in the future.

2 should be easy.

If 1 did pre-exist and isn’t a very quick fix, another ticket is fine and we can merge this, but as that comes up as a notified error back to the GUI it will have to be a priority fix anyway, otherwise we start risking error-blindness if there are too many standard errors that can be ignored.

**Comment by AdrianPotter** [comment]
Confirmed that I'm getting error reported on SampleChanger and LinearSampleChanger on reload config. I've added the ticket as 1468 as requested.

I totally see the point about excess information in the support submodule. From a dev perspective, it's useful having something I can just grab to have a configured device set up. Otherwise I have to look at something like LARMOR and derive a config from one of theirs. I'm not familiar with Galils to set all the config up from scratch from something blank. I entirely appreciate it risks being rendered incompatible by any breaking changes but I'm not sure if there's a better place to put it. Shall I trim it some more or delete it entirely?

**Comment by KathrynBaker** [comment]
Thank you for the ticket.

In the spirit of moving this forward, I’ve merged as is (and moved this to complete), with the full config in place (we can remove it later), but it is something we should probably address and agree on. One of those places where documentation is useful, so that information is not in the head of a couple of people but is shared, as the instructions should be fairly generic for testing (or am I the only one who goes in and reduces or expands example files when testing?). We can use the config repo for things like this as well I should think. The one thing that it could be worth doing, as we move forward with new functionality, or as we change existing functionality, is to write the instructions along the lines of:

Add a line like `some bits to add` to specificFile.cmd (which can be found in specificDir)
Put a file called newFile.blah in specificDir with the following code:

 ```A few lines of example code
Just to show what could be done
Not necessarily code for an instrument
But it should run```

But we should probably put these instructions in the wiki ready for inclusion in documentation for instrument scientist/technician/developer type people.

**Comment by ChrisM-S** [comment]
Should we be reviewing code for poetic content too… ☺
